service/src/test/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureNetworkInterfaceStepUnitTest.java

```diff
@@ -0,0 +1,182 @@
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.mockito.Mockito.*;
+import static org.mockito.Mockito.verify;
+
+@Tag("unit")
+@ExtendWith(MockitoExtension.class)
+public class CreateAzureNetworkInterfaceStepUnitTest {
+
+  @Mock AzureConfiguration azureConfiguration;
+  @Mock CrlService crlService;
+  @Mock ControlledAzureVmResource resource;
+  @Mock SamService samService;
+  @Mock WorkspaceService mockWorkspaceService;
+  @Mock ResourceDao resourceDao;
+  @Mock LandingZoneApiDispatch landingZoneApiDispatch;
+  @Mock NetworkManager networkManager;
+
+
+
+  @BeforeEach
+  void setup() {
+
+  }
+
+  @Test
+  void undo_HappyPath() throws Exception {
+    AzureCloudContext azureCloudContext = mock();
+    var resourceGroupId = "test-resource-group-id";
+    when(azureCloudContext.getAzureResourceGroupId()).thenReturn(resourceGroupId);
+    FlightMap workingMap = mock();
+    when(workingMap.get(WorkspaceFlightMapKeys.ControlledResourceKeys.AZURE_CLOUD_CONTEXT,
+        AzureCloudContext.class)).thenReturn(azureCloudContext);
+    var resourceName = "test-vm-name";
+    //when(resource.getVmName()).thenReturn(resourceName);
+    var networkInterfaceName =  String.format("nic-%s", resourceName);
+    when(workingMap.get(AzureVmHelper.WORKING_MAP_NETWORK_INTERFACE_KEY, String.class))
+        .thenReturn(networkInterfaceName);
+
+    FlightContext context = mock();
+
+    when(context.getWorkingMap()).thenReturn(workingMap);
+    ComputeManager computeManager = mock();
+    when(crlService.getComputeManager(azureCloudContext, azureConfiguration)).thenReturn(computeManager);
+    when(computeManager.networkManager()).thenReturn(networkManager);
+    NetworkInterfaces networkInterfaces = mock();
+    when(networkManager.networkInterfaces()).thenReturn(networkInterfaces);
+
+    doNothing().when(networkInterfaces).deleteByResourceGroup(resourceGroupId, networkInterfaceName);
+
+    var step = new CreateAzureNetworkInterfaceStep(
+        azureConfiguration,
+        crlService,
+        resource,
+        resourceDao,
+        landingZoneApiDispatch,
+        samService,
+        mockWorkspaceService);
+    assertThat(step.undoStep(context), equalTo(StepResult.getStepResultSuccess()));
+
+    verify(networkInterfaces).deleteByResourceGroup(resourceGroupId, networkInterfaceName);
+  }
+
+
+  @Test
+  void undo_SucceedsWhenResourceNotFound() throws Exception {
+    AzureCloudContext azureCloudContext = mock();
+    var resourceGroupId = "test-resource-group-id";
+    when(azureCloudContext.getAzureResourceGroupId()).thenReturn(resourceGroupId);
+    FlightMap workingMap = mock();
+    when(workingMap.get(WorkspaceFlightMapKeys.ControlledResourceKeys.AZURE_CLOUD_CONTEXT,
+        AzureCloudContext.class)).thenReturn(azureCloudContext);
+    var resourceName = "test-vm-name";
+    //when(resource.getVmName()).thenReturn(resourceName);
```

Comment:
comment

Reply:
Can you move some of this common setup to the `setup` method?

Reply:
I moved some of it.

Some parts can't go there because of the strict stubbing default, though.